---
title: "Clustering of differentially expressed genes (RNAseq)"
output: html_document
---

```{r simulation_options, include=FALSE}
library(knitr)
library(rmarkdown)
library(report)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(ggplot2)
library(reshape2)
opts_chunk$set(fig.width = 10, fig.height = 10, dpi = 72,
               warning = FALSE, message = FALSE, prompt = TRUE, cache = FALSE)
set.seed(123)
```
last changed: `r Sys.Date()`

---

Hierarchical Clustering
===

Load the `rlog` transformed DESeq data set and construct the required data 
frames. The function `assay` returns the values and `colData` returns the 
information about the experimental design.
```{r data1}
rlog.heat <- readRDS(file = "./../Significance_testing/Data/rlog_RNAseq_sig.rds")
head(rlog.heat)

m.RNAseq <- assay(rlog.heat)
head(m.RNAseq)

info.RNAseq <- data.frame(colData(rlog.heat))
paged_table(info.RNAseq)
```

---

Next this data can be used to perform hierarchical clustering, which is 
implemented in `pheatmap`. `colorRampPallette` was used to change the annotation 
color.
```{r hier_clus}
col <- colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100)

heat.hieracial <- pheatmap(m.RNAseq, 
                  cluster_cols = F, 
                  show_rownames = F, 
                  scale = "row", 
                  color = col,
                  annotation_col = data.frame(time = info.RNAseq[, "Time"], 
                                              row.names = rownames(info.RNAseq)))
```

---

Kmeans Clustering
===

To get a better separation of the data, kmeans clustering was performed. For 
this, the data was scaled to Z-scores using the function `scale`.
```{r data2}
norm.m.RNAseq <- t(scale(t(m.RNAseq)))
```

---

Next, `kmeans` was used to perform the clustering and a factor indicating the 
clusters was added to each of the scaled matrices. The clustering for
`centers = 5` (5 cluster) yielded the best result, as more clusters divided 
data with a similar course.
```{r kmeans}
clus.kmeans <- kmeans(norm.m.RNAseq, 4)
lapply(clus.kmeans, head)

saveRDS(clus.kmeans, "./Data/kmeans.rds")

m.kmeans <- cbind(norm.m.RNAseq, cluster = as.factor(clus.kmeans$cluster))
m.kmeans <- m.kmeans[order(m.kmeans[, "cluster"]), ]
```

---

To visualize the clustering a heatmap was plotted using `pheatmap`. The functions 
`colorRampPalette` and `brewer.pal` were used to generate the color for the 
annotations and Z-values, respectively.
```{r heat, fig}
col <- colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100)
col2 <- brewer.pal(n = 8, name = "Dark2")
clus.col <- c("1" = col2[1], "2" = col2[2], "3" = col2[3], "4" = col2[4])
ann.col2 <- list(cluster = clus.col[1:4])

heat.kmeans.RNAseq <- pheatmap(m.kmeans[ ,-ncol(m.kmeans)],
                              cluster_rows = F,
                              cluster_cols = F,
                              show_rownames = F, 
                              color = col,
                              annotation_col = data.frame(time = info.RNAseq[, "Time"], row.names = rownames(info.RNAseq)),
                              annotation_row = data.frame(cluster = as.factor(m.kmeans[, "cluster"])),
                              annotation_colors = ann.col2,
                              main = "Gene expression")
```

---

Lastly, the cluster centroids (Z-scores) were plotted separately to visualize 
the change in H3K27ac over time between the two constructs.
```{r Z_cen}
kmeans <- melt(clus.kmeans$centers)
ind <- match(kmeans$Var2, rownames(info.RNAseq))
data.kmeans <- cbind(kmeans, info.RNAseq[ind, "Time"])
colnames(data.kmeans) <- c("cluster", "name", "Z_score", "time")
paged_table(data.kmeans)
  
ggplot(data.kmeans) + 
    geom_point(aes(x = time, y = Z_score), size = 1) +
    stat_summary(aes(x = time, y = Z_score), fun = mean, geom = "line", linewidth = 1) +
    facet_grid(rows = data.kmeans$cluster) +
    theme_bw() +
    labs(title = "Cluster centroids", y = "z-scores")
```

---

References - packages
---------------------

```{r packages, echo=FALSE, results='asis'}
cite_packages()
```

---

Session Information
-------------------

The following versions of R and R packages were used to generate the report
above:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```